File: internal/handlers/dns.go
package handlers

import (
	"net/http"

	"github.com/labstack/echo/v4"

	"codedock.dev/codedock/internal/models"
	"codedock.dev/codedock/internal/services"
	"codedock.dev/codedock/internal/utils"
)

type DNSHandler struct {
	dnsService *services.DNSService
}

func NewDNSHandler(dnsService *services.DNSService) *DNSHandler {
	return &DNSHandler{dnsService: dnsService}
}

func (h *DNSHandler) Create(c echo.Context) error {
	var req models.CreateDNSRecordRequest
	if err := c.Bind(&req); err != nil {
		return utils.Error(c, http.StatusBadRequest, "invalid payload")
	}

	record, err := h.dnsService.CreateRecord(c.Request().Context(), &req)
	if err != nil {
		return utils.Error(c, http.StatusInternalServerError, err.Error())
	}

	return utils.Created(c, "DNS record created successfully", record)
}

func (h *DNSHandler) List(c echo.Context) error {
	domain := c.QueryParam("domain")

	records, err := h.dnsService.ListByDomain(c.Request().Context(), domain)
	if err != nil {
		return utils.Error(c, http.StatusInternalServerError, err.Error())
	}

	return utils.Success(c, "DNS records fetched successfully", records)
}

func (h *DNSHandler) Update(c echo.Context) error {
	id := c.Param("id")
	if id == "" {
		return utils.Error(c, http.StatusBadRequest, "missing record id")
	}

	var req models.UpdateDNSRecordRequest
	if err := c.Bind(&req); err != nil {
		return utils.Error(c, http.StatusBadRequest, "invalid payload")
	}

	record, err := h.dnsService.UpdateRecord(c.Request().Context(), id, &req)
	if err != nil {
		return utils.Error(c, http.StatusInternalServerError, err.Error())
	}

	return utils.Success(c, "DNS record updated successfully", record)
}

func (h *DNSHandler) Delete(c echo.Context) error {
	id := c.Param("id")
	if id == "" {
		return utils.Error(c, http.StatusBadRequest, "missing record id")
	}

	if err := h.dnsService.DeleteRecord(c.Request().Context(), id); err != nil {
		return utils.Error(c, http.StatusInternalServerError, err.Error())
	}

	return c.NoContent(http.StatusNoContent)
}

File: apps/dashboard/src/services/dns.ts
import type { BaseResponse } from '#/interfaces/base';
import type { CreateDNSRecordRequest, DNSRecord, UpdateDNSRecordRequest } from '#/interfaces/dns';
import { apiClient } from '#/lib/apiClient';
import { handleApiError } from '#/lib/error';

export const dnsService = {
  list: async (): Promise<BaseResponse<DNSRecord[]>> => {
    try {
      return await apiClient.get<BaseResponse<DNSRecord[]>>(`/dns`);
    } catch (error) {
      throw handleApiError(error);
    }
  },

  create: async (payload: CreateDNSRecordRequest): Promise<BaseResponse<DNSRecord>> => {
    try {
      return await apiClient.post<BaseResponse<DNSRecord>>(`/dns`, payload);
    } catch (error) {
      throw handleApiError(error);
    }
  },

  update: async (id: string, payload: UpdateDNSRecordRequest): Promise<BaseResponse<DNSRecord>> => {
    try {
      return await apiClient.put<BaseResponse<DNSRecord>>(`/dns/${id}`, payload);
    } catch (error) {
      throw handleApiError(error);
    }
  },

  delete: async (id: string): Promise<void> => {
    try {
      await apiClient.delete(`/dns/${id}`);
    } catch (error) {
      throw handleApiError(error);
    }
  },
};
